File: src/harness_core/usage_ledger.py
"""Usage Ledger and Cost-of-Pass schema and validation helpers.

Records per-eval-row token/cost/retry/tool-call data and aggregates
by cost_of_pass_group for cost-efficiency comparison.
"""

from __future__ import annotations

import json
import re
from collections import defaultdict
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any, Dict, List, Optional, Sequence, Tuple


# ---------------------------------------------------------------------------
# Schema version
# ---------------------------------------------------------------------------

SCHEMA_VERSION = "usage_ledger.v1"

REQUIRED_FIELDS: Sequence[str] = (
    "schema_version",
    "run_id",
    "case_id",
    "input_tokens",
    "output_tokens",
    "cached_tokens",
    "request_count",
    "tool_call_count",
    "retry_count",
    "wall_clock_ms",
    "estimated_cost",
    "pass",
    "cost_of_pass_group",
    "model_profile_id",
    "context_pack_id",
)

# cost_of_pass_group format: <eval_suite>/<task_family>/<variant_family>/<success_criterion>
COST_OF_PASS_GROUP_PATTERN = re.compile(
    r"^[a-zA-Z0-9_\-]+/[a-zA-Z0-9_\-]+/[a-zA-Z0-9_\-]+/[a-zA-Z0-9_\-]+$"
)


# ---------------------------------------------------------------------------
# Data class
# ---------------------------------------------------------------------------

@dataclass
class UsageLedgerRow:
    """A single usage ledger record."""

    run_id: str
    case_id: str
    input_tokens: int
    output_tokens: int
    cached_tokens: int
    request_count: int
    tool_call_count: int
    retry_count: int
    wall_clock_ms: int
    estimated_cost: float
    pass_: bool
    cost_of_pass_group: str
    model_profile_id: str
    context_pack_id: str
    schema_version: str = SCHEMA_VERSION

    def to_dict(self) -> Dict[str, Any]:
        d = asdict(self)
        d["pass"] = d.pop("pass_")
        return d

    def to_json(self, **kw: Any) -> str:
        return json.dumps(self.to_dict(), **kw)


@dataclass
class CostOfPassAggregate:
    """Aggregated cost-of-pass for a single cost_of_pass_group."""

    cost_of_pass_group: str
    total_estimated_cost: float
    success_count: int
    failure_count: int
    total_count: int
    cost_of_pass: Optional[float]  # None when success_count == 0

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


@dataclass
class ComparisonResult:
    """Result of comparing two cost-of-pass groups."""

    group_a: str
    group_b: str
    aggregate_a: CostOfPassAggregate
    aggregate_b: CostOfPassAggregate
    valid: bool
    reason: str
    cost_delta: Optional[float] = None
    relative_change_pct: Optional[float] = None

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------

def validate_usage_ledger_row(data: Dict[str, Any]) -> List[str]:
    """Validate a usage_ledger row dict. Returns list of violations."""
    violations: List[str] = []

    # 1. Required fields
    for f in REQUIRED_FIELDS:
        if f not in data:
            violations.append(f"missing required field: {f}")
    if violations:
        return violations

    # 2. schema_version
    if data["schema_version"] != SCHEMA_VERSION:
        violations.append(f"schema_version must be {SCHEMA_VERSION}, got {data['schema_version']!r}")

    # 3. pass must be bool
    if not isinstance(data["pass"], bool):
        violations.append(f"pass must be a bool, got {type(data['pass']).__name__}")

    # 4. Non-negative integer fields
    int_fields = ("input_tokens", "output_tokens", "cached_tokens",
                  "request_count", "tool_call_count", "retry_count", "wall_clock_ms")
    for f in int_fields:
        if not isinstance(data[f], int) or data[f] < 0:
            violations.append(f"{f} must be a non-negative integer, got {data[f]!r}")

    # 5. estimated_cost non-negative
    if not isinstance(data["estimated_cost"], (int, float)) or data["estimated_cost"] < 0:
        violations.append(f"estimated_cost must be a non-negative number, got {data['estimated_cost']!r}")

    # 6. cached_tokens <= input_tokens
    if isinstance(data["cached_tokens"], int) and isinstance(data["input_tokens"], int):
        if data["cached_tokens"] > data["input_tokens"]:
            violations.append(
                f"cached_tokens ({data['cached_tokens']}) must not exceed "
                f"input_tokens ({data['input_tokens']})"
            )

    # 7. cost_of_pass_group format
    if not COST_OF_PASS_GROUP_PATTERN.match(data["cost_of_pass_group"]):
        violations.append(
            f"cost_of_pass_group {data['cost_of_pass_group']!r} does not match "
            "format <eval_suite>/<task_family>/<variant_family>/<success_criterion>"
        )

    # 8. model_profile_id and context_pack_id: allow empty string or null
    for f in ("model_profile_id", "context_pack_id"):
        val = data[f]
        if val is not None and val != "" and not isinstance(val, str):
            violations.append(f"{f} must be a string, null, or empty string")

    return violations


def is_valid_cost_of_pass_group(group: str) -> bool:
    """Check if a cost_of_pass_group matches the required four-segment format."""
    return bool(COST_OF_PASS_GROUP_PATTERN.match(group))


def parse_cost_of_pass_group(group: str) -> Tuple[str, str, str, str]:
    """Parse a cost_of_pass_group into its four segments.

    Returns (eval_suite, task_family, variant_family, success_criterion).
    Raises ValueError if format is invalid.
    """
    if not COST_OF_PASS_GROUP_PATTERN.match(group):
        raise ValueError(
            f"cost_of_pass_group {group!r} does not match "
            "<eval_suite>/<task_family>/<variant_family>/<success_criterion>"
        )
    parts = group.split("/")
    return parts[0], parts[1], parts[2], parts[3]


# ---------------------------------------------------------------------------
# Aggregation
# ---------------------------------------------------------------------------

def aggregate_cost_of_pass(rows: List[Dict[str, Any]]) -> CostOfPassAggregate:
    """Aggregate usage ledger rows into a single cost-of-pass metric.

    cost_of_pass = total_estimated_cost / success_count
    If success_count == 0, cost_of_pass is None and this is a failure.
    """
    if not rows:
        return CostOfPassAggregate(
            cost_of_pass_group="",
            total_estimated_cost=0.0,
            success_count=0,
            failure_count=0,
            total_count=0,
            cost_of_pass=None,
        )

    group = rows[0].get("cost_of_pass_group", "")
    total_cost = 0.0
    success = 0
    failure = 0

    for row in rows:
        total_cost += row.get("estimated_cost", 0)
        if row.get("pass") is True:
            success += 1
        else:
            failure += 1

    total = success + failure
    cop = total_cost / success if success > 0 else None

    return CostOfPassAggregate(
        cost_of_pass_group=group,
        total_estimated_cost=total_cost,
        success_count=success,
        failure_count=failure,
        total_count=total,
        cost_of_pass=cop,
    )


def group_usage_rows(rows: List[Dict[str, Any]]) -> Dict[str, List[Dict[str, Any]]]:
    """Group usage ledger rows by cost_of_pass_group."""
    groups: Dict[str, List[Dict[str, Any]]] = defaultdict(list)
    for row in rows:
        groups[row["cost_of_pass_group"]].append(row)
    return dict(groups)


def compare_cost_groups(
    before_rows: List[Dict[str, Any]],
    after_rows: List[Dict[str, Any]],
) -> ComparisonResult:
    """Compare cost-of-pass between two sets of rows (before/after).

    Both must be in the same cost_of_pass_group for a valid comparison.
    """
    before_agg = aggregate_cost_of_pass(before_rows)
    after_agg = aggregate_cost_of_pass(after_rows)

    # Check same group
    if before_agg.cost_of_pass_group != after_agg.cost_of_pass_group:
        return ComparisonResult(
            group_a=before_agg.cost_of_pass_group,
            group_b=after_agg.cost_of_pass_group,
            aggregate_a=before_agg,
            aggregate_b=after_agg,
            valid=False,
            reason="cannot compare different cost_of_pass_groups directly",
        )

    # Check both have defined cost_of_pass
    if before_agg.cost_of_pass is None or after_agg.cost_of_pass is None:
        return ComparisonResult(
            group_a=before_agg.cost_of_pass_group,
            group_b=after_agg.cost_of_pass_group,
            aggregate_a=before_agg,
            aggregate_b=after_agg,
            valid=False,
            reason="cost_of_pass undefined for one or both groups (success_count=0)",
        )

    delta = after_agg.cost_of_pass - before_agg.cost_of_pass
    pct = (delta / before_agg.cost_of_pass * 100) if before_agg.cost_of_pass != 0 else None

    return ComparisonResult(
        group_a=before_agg.cost_of_pass_group,
        group_b=after_agg.cost_of_pass_group,
        aggregate_a=before_agg,
        aggregate_b=after_agg,
        valid=True,
        reason="same group, both have defined cost_of_pass",
        cost_delta=delta,
        relative_change_pct=pct,
    )


def detect_invalid_cost_comparison(
    group_a_rows: List[Dict[str, Any]],
    group_b_rows: List[Dict[str, Any]],
) -> Tuple[bool, str]:
    """Detect if comparing two groups would be an invalid cost comparison.

    Returns (is_invalid, reason).
    """
    agg_a = aggregate_cost_of_pass(group_a_rows)
    agg_b = aggregate_cost_of_pass(group_b_rows)

    if agg_a.cost_of_pass_group != agg_b.cost_of_pass_group:
        return True, (
            f"groups {agg_a.cost_of_pass_group!r} and {agg_b.cost_of_pass_group!r} "
            "are different; direct cost comparison is invalid"
        )

    if agg_a.cost_of_pass is None or agg_b.cost_of_pass is None:
        return True, "one or both groups have undefined cost_of_pass (success_count=0)"

    return False, "comparison is valid"


# ---------------------------------------------------------------------------
# Fixture loading
# ---------------------------------------------------------------------------

def load_fixture(path: Path) -> Dict[str, Any]:
    with open(path, "r") as f:
        return json.load(f)


def load_usage_ledger_fixture(path: Path) -> Tuple[Dict[str, Any], List[str]]:
    """Load a usage ledger fixture and validate it."""
    data = load_fixture(path)
    return data, validate_usage_ledger_row(data)


def load_all_fixtures(fixture_dir: Path) -> List[Tuple[str, Dict[str, Any], List[str]]]:
    """Load and validate every .json fixture in a directory."""
    results = []
    for p in sorted(fixture_dir.glob("*.json")):
        data = load_fixture(p)
        violations = validate_usage_ledger_row(data)
        results.append((p.name, data, violations))
    return results

File: engine/src/app_layer/usage_ledger.rs
use std::collections::HashMap;

use regex::Regex;
use serde::{Deserialize, Serialize};
use std::sync::LazyLock;

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

pub const SCHEMA_VERSION: &str = "usage_ledger.v1";

pub const REQUIRED_FIELDS: &[&str] = &[
    "schema_version",
    "run_id",
    "case_id",
    "input_tokens",
    "output_tokens",
    "cached_tokens",
    "request_count",
    "tool_call_count",
    "retry_count",
    "wall_clock_ms",
    "estimated_cost",
    "pass",
    "cost_of_pass_group",
    "model_profile_id",
    "context_pack_id",
];

static COST_OF_PASS_GROUP_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^[a-zA-Z0-9_\-]+/[a-zA-Z0-9_\-]+/[a-zA-Z0-9_\-]+/[a-zA-Z0-9_\-]+$").unwrap()
});

// ---------------------------------------------------------------------------
// Data structs
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct UsageLedgerRow {
    pub run_id: String,
    pub case_id: String,
    pub input_tokens: i64,
    pub output_tokens: i64,
    pub cached_tokens: i64,
    pub request_count: i64,
    pub tool_call_count: i64,
    pub retry_count: i64,
    pub wall_clock_ms: i64,
    pub estimated_cost: f64,
    #[serde(rename = "pass")]
    pub pass_: bool,
    pub cost_of_pass_group: String,
    pub model_profile_id: String,
    pub context_pack_id: String,
    pub schema_version: String,
}

impl Default for UsageLedgerRow {
    fn default() -> Self {
        Self {
            run_id: String::new(),
            case_id: String::new(),
            input_tokens: 0,
            output_tokens: 0,
            cached_tokens: 0,
            request_count: 0,
            tool_call_count: 0,
            retry_count: 0,
            wall_clock_ms: 0,
            estimated_cost: 0.0,
            pass_: false,
            cost_of_pass_group: String::new(),
            model_profile_id: String::new(),
            context_pack_id: String::new(),
            schema_version: SCHEMA_VERSION.to_string(),
        }
    }
}

impl UsageLedgerRow {
    pub fn to_value(&self) -> serde_json::Value {
        serde_json::to_value(self).unwrap()
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct CostOfPassAggregate {
    pub cost_of_pass_group: String,
    pub total_estimated_cost: f64,
    pub success_count: i64,
    pub failure_count: i64,
    pub total_count: i64,
    pub cost_of_pass: Option<f64>,
}

impl Default for CostOfPassAggregate {
    fn default() -> Self {
        Self {
            cost_of_pass_group: String::new(),
            total_estimated_cost: 0.0,
            success_count: 0,
            failure_count: 0,
            total_count: 0,
            cost_of_pass: None,
        }
    }
}

impl CostOfPassAggregate {
    pub fn to_value(&self) -> serde_json::Value {
        serde_json::to_value(self).unwrap()
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct ComparisonResult {
    pub group_a: String,
    pub group_b: String,
    pub aggregate_a: CostOfPassAggregate,
    pub aggregate_b: CostOfPassAggregate,
    pub valid: bool,
    pub reason: String,
    pub cost_delta: Option<f64>,
    pub relative_change_pct: Option<f64>,
}

impl Default for ComparisonResult {
    fn default() -> Self {
        Self {
            group_a: String::new(),
            group_b: String::new(),
            aggregate_a: CostOfPassAggregate::default(),
            aggregate_b: CostOfPassAggregate::default(),
            valid: false,
            reason: String::new(),
            cost_delta: None,
            relative_change_pct: None,
        }
    }
}

impl ComparisonResult {
    pub fn to_value(&self) -> serde_json::Value {
        serde_json::to_value(self).unwrap()
    }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

pub fn validate_usage_ledger_row(data: &serde_json::Value) -> Vec<String> {
    let mut violations = Vec::new();

    let object = match data.as_object() {
        Some(o) => o,
        None => {
            violations.push("usage_ledger_row must be a JSON object".to_string());
            return violations;
        }
    };

    for field in REQUIRED_FIELDS {
        if !object.contains_key(*field) {
            violations.push(format!("missing required field: {}", field));
        }
    }
    if !violations.is_empty() {
        return violations;
    }

    let sv = object["schema_version"].as_str().unwrap_or("");
    if sv != SCHEMA_VERSION {
        violations.push(format!(
            "schema_version must be {}, got {:?}",
            SCHEMA_VERSION, sv
        ));
    }

    if !object["pass"].is_boolean() {
        violations.push(format!(
            "pass must be a bool, got {}",
            json_type_name(&object["pass"])
        ));
    }

    let int_fields = [
        "input_tokens",
        "output_tokens",
        "cached_tokens",
        "request_count",
        "tool_call_count",
        "retry_count",
        "wall_clock_ms",
    ];
    for field in &int_fields {
        match object[*field].as_i64() {
            Some(v) if v >= 0 => {}
            _ => violations.push(format!(
                "{} must be a non-negative integer, got {}",
                field, object[*field]
            )),
        }
    }

    match object["estimated_cost"].as_f64() {
        Some(v) if v >= 0.0 => {}
        _ => violations.push(format!(
            "estimated_cost must be a non-negative number, got {}",
            object["estimated_cost"]
        )),
    }

    if let (Some(cached), Some(input)) = (
        object["cached_tokens"].as_i64(),
        object["input_tokens"].as_i64(),
    ) {
        if cached > input {
            violations.push(format!(
                "cached_tokens ({}) must not exceed input_tokens ({})",
                cached, input
            ));
        }
    }

    let group = object["cost_of_pass_group"].as_str().unwrap_or("");
    if !COST_OF_PASS_GROUP_PATTERN.is_match(group) {
        violations.push(format!(
            "cost_of_pass_group {:?} does not match format <eval_suite>/<task_family>/<variant_family>/<success_criterion>",
            group
        ));
    }

    for field in &["model_profile_id", "context_pack_id"] {
        let val = &object[*field];
        if !val.is_null() && !val.is_string() {
            violations.push(format!("{} must be a string, null, or empty string", field));
        }
    }

    violations
}

fn json_type_name(v: &serde_json::Value) -> &'static str {
    match v {
        serde_json::Value::Null => "null",
        serde_json::Value::Bool(_) => "bool",
        serde_json::Value::Number(_) => "number",
        serde_json::Value::String(_) => "string",
        serde_json::Value::Array(_) => "array",
        serde_json::Value::Object(_) => "object",
    }
}

// ---------------------------------------------------------------------------
// Group format helpers
// ---------------------------------------------------------------------------

pub fn is_valid_cost_of_pass_group(group: &str) -> bool {
    COST_OF_PASS_GROUP_PATTERN.is_match(group)
}

pub fn parse_cost_of_pass_group(group: &str) -> Result<(String, String, String, String), String> {
    if !COST_OF_PASS_GROUP_PATTERN.is_match(group) {
        return Err(format!(
            "cost_of_pass_group {:?} does not match <eval_suite>/<task_family>/<variant_family>/<success_criterion>",
            group
        ));
    }
    let parts: Vec<&str> = group.split('/').collect();
    Ok((
        parts[0].to_string(),
        parts[1].to_string(),
        parts[2].to_string(),
        parts[3].to_string(),
    ))
}

// ---------------------------------------------------------------------------
// Aggregation
// ---------------------------------------------------------------------------

pub fn aggregate_cost_of_pass(rows: &[serde_json::Value]) -> CostOfPassAggregate {
    if rows.is_empty() {
        return CostOfPassAggregate::default();
    }

    let group = rows[0]
        .get("cost_of_pass_group")
        .and_then(serde_json::Value::as_str)
        .unwrap_or("")
        .to_string();

    let mut total_cost = 0.0_f64;
    let mut success: i64 = 0;
    let mut failure: i64 = 0;

    for row in rows {
        total_cost += row
            .get("estimated_cost")
            .and_then(serde_json::Value::as_f64)
            .unwrap_or(0.0);
        if row.get("pass").and_then(serde_json::Value::as_bool) == Some(true) {
            success += 1;
        } else {
            failure += 1;
        }
    }

    let total = success + failure;
    let cop = if success > 0 {
        Some(total_cost / success as f64)
    } else {
        None
    };

    CostOfPassAggregate {
        cost_of_pass_group: group,
        total_estimated_cost: total_cost,
        success_count: success,
        failure_count: failure,
        total_count: total,
        cost_of_pass: cop,
    }
}

pub fn group_usage_rows(rows: &[serde_json::Value]) -> HashMap<String, Vec<serde_json::Value>> {
    let mut groups: HashMap<String, Vec<serde_json::Value>> = HashMap::new();
    for row in rows {
        let group = row
            .get("cost_of_pass_group")
            .and_then(serde_json::Value::as_str)
            .unwrap_or("")
            .to_string();
        groups.entry(group).or_default().push(row.clone());
    }
    groups
}

pub fn compare_cost_groups(
    before_rows: &[serde_json::Value],
    after_rows: &[serde_json::Value],
) -> ComparisonResult {
    let before_agg = aggregate_cost_of_pass(before_rows);
    let after_agg = aggregate_cost_of_pass(after_rows);

    if before_agg.cost_of_pass_group != after_agg.cost_of_pass_group {
        return ComparisonResult {
            group_a: before_agg.cost_of_pass_group.clone(),
            group_b: after_agg.cost_of_pass_group.clone(),
            aggregate_a: before_agg,
            aggregate_b: after_agg,
            valid: false,
            reason: "cannot compare different cost_of_pass_groups directly".to_string(),
            cost_delta: None,
            relative_change_pct: None,
        };
    }

    if before_agg.cost_of_pass.is_none() || after_agg.cost_of_pass.is_none() {
        return ComparisonResult {
            group_a: before_agg.cost_of_pass_group.clone(),
            group_b: after_agg.cost_of_pass_group.clone(),
            aggregate_a: before_agg,
            aggregate_b: after_agg,
            valid: false,
            reason: "cost_of_pass undefined for one or both groups (success_count=0)".to_string(),
            cost_delta: None,
            relative_change_pct: None,
        };
    }

    let before_cop = before_agg.cost_of_pass.unwrap();
    let after_cop = after_agg.cost_of_pass.unwrap();
    let delta = after_cop - before_cop;
    let pct = if before_cop != 0.0 {
        Some(delta / before_cop * 100.0)
    } else {
        None
    };

    ComparisonResult {
        group_a: before_agg.cost_of_pass_group.clone(),
        group_b: after_agg.cost_of_pass_group.clone(),
        aggregate_a: before_agg,
        aggregate_b: after_agg,
        valid: true,
        reason: "same group, both have defined cost_of_pass".to_string(),
        cost_delta: Some(delta),
        relative_change_pct: pct,
    }
}

pub fn detect_invalid_cost_comparison(
    group_a_rows: &[serde_json::Value],
    group_b_rows: &[serde_json::Value],
) -> (bool, String) {
    let agg_a = aggregate_cost_of_pass(group_a_rows);
    let agg_b = aggregate_cost_of_pass(group_b_rows);

    if agg_a.cost_of_pass_group != agg_b.cost_of_pass_group {
        return (
            true,
            format!(
                "groups {:?} and {:?} are different; direct cost comparison is invalid",
                agg_a.cost_of_pass_group, agg_b.cost_of_pass_group
            ),
        );
    }

    if agg_a.cost_of_pass.is_none() || agg_b.cost_of_pass.is_none() {
        return (
            true,
            "one or both groups have undefined cost_of_pass (success_count=0)".to_string(),
        );
    }

    (false, "comparison is valid".to_string())
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    const GROUP: &str = "eval_suite/task_family/variant_family/success_criterion";

    fn valid_row(pass: bool, cost: f64, group: &str) -> serde_json::Value {
        json!({
            "schema_version": SCHEMA_VERSION,
            "run_id": "run-1",
            "case_id": "case-1",
            "input_tokens": 100,
            "output_tokens": 50,
            "cached_tokens": 10,
            "request_count": 1,
            "tool_call_count": 0,
            "retry_count": 0,
            "wall_clock_ms": 500,
            "estimated_cost": cost,
            "pass": pass,
            "cost_of_pass_group": group,
            "model_profile_id": "mimo-v2",
            "context_pack_id": "cp-001"
        })
    }

    #[test]
    fn test_valid_row_passes() {
        let violations = validate_usage_ledger_row(&valid_row(true, 0.05, GROUP));
        assert!(
            violations.is_empty(),
            "expected no violations, got {:?}",
            violations
        );
    }

    #[test]
    fn test_missing_required_field() {
        let mut data = valid_row(true, 0.01, GROUP);
        data.as_object_mut().unwrap().remove("run_id");
        let violations = validate_usage_ledger_row(&data);
        assert!(violations.iter().any(|v| v.contains("run_id")));
    }

    #[test]
    fn test_negative_input_tokens() {
        let mut data = valid_row(true, 0.01, GROUP);
        data["input_tokens"] = json!(-5);
        let violations = validate_usage_ledger_row(&data);
        assert!(violations
            .iter()
            .any(|v| v.contains("input_tokens") && v.contains("non-negative")));
    }

    #[test]
    fn test_cached_exceeds_input() {
        let mut data = valid_row(true, 0.01, GROUP);
        data["cached_tokens"] = json!(200);
        data["input_tokens"] = json!(100);
        let violations = validate_usage_ledger_row(&data);
        assert!(violations
            .iter()
            .any(|v| v.contains("cached_tokens") && v.contains("must not exceed")));
    }

    #[test]
    fn test_bad_cost_of_pass_group_format() {
        let mut data = valid_row(true, 0.01, GROUP);
        data["cost_of_pass_group"] = json!("bad_format");
        let violations = validate_usage_ledger_row(&data);
        assert!(violations.iter().any(|v| v.contains("cost_of_pass_group")));
    }

    #[test]
    fn test_is_valid_cost_of_pass_group() {
        assert!(is_valid_cost_of_pass_group(GROUP));
        assert!(!is_valid_cost_of_pass_group("bad"));
        assert!(!is_valid_cost_of_pass_group("a/b/c"));
        assert!(!is_valid_cost_of_pass_group("a/b/c/d/e"));
    }

    #[test]
    fn test_parse_cost_of_pass_group() {
        let (a, b, c, d) = parse_cost_of_pass_group(GROUP).unwrap();
        assert_eq!(a, "eval_suite");
        assert_eq!(b, "task_family");
        assert_eq!(c, "variant_family");
        assert_eq!(d, "success_criterion");

        assert!(parse_cost_of_pass_group("bad").is_err());
    }

    #[test]
    fn test_aggregate_cost_of_pass_basic() {
        let rows = vec![
            valid_row(true, 0.10, GROUP),
            valid_row(true, 0.20, GROUP),
            valid_row(false, 0.05, GROUP),
        ];
        let agg = aggregate_cost_of_pass(&rows);
        assert_eq!(agg.success_count, 2);
        assert_eq!(agg.failure_count, 1);
        assert_eq!(agg.total_count, 3);
        assert!((agg.total_estimated_cost - 0.35).abs() < 1e-10);
        assert!(agg.cost_of_pass.is_some());
        let cop = agg.cost_of_pass.unwrap();
        assert!((cop - 0.175).abs() < 1e-10);
    }

    #[test]
    fn test_aggregate_empty_rows() {
        let agg = aggregate_cost_of_pass(&[]);
        assert_eq!(agg.total_count, 0);
        assert!(agg.cost_of_pass.is_none());
    }

    #[test]
    fn test_aggregate_all_failures() {
        let rows = vec![valid_row(false, 0.10, GROUP), valid_row(false, 0.20, GROUP)];
        let agg = aggregate_cost_of_pass(&rows);
        assert_eq!(agg.success_count, 0);
        assert_eq!(agg.failure_count, 2);
        assert!(agg.cost_of_pass.is_none());
    }

    #[test]
    fn test_group_usage_rows() {
        let group_a = "suite/fam/variant/crit_a";
        let group_b = "suite/fam/variant/crit_b";
        let rows = vec![
            valid_row(true, 0.1, group_a),
            valid_row(true, 0.2, group_a),
            valid_row(false, 0.3, group_b),
        ];
        let groups = group_usage_rows(&rows);
        assert_eq!(groups.len(), 2);
        assert_eq!(groups[group_a].len(), 2);
        assert_eq!(groups[group_b].len(), 1);
    }

    #[test]
    fn test_compare_cost_groups_valid() {
        let before = vec![valid_row(true, 0.10, GROUP), valid_row(true, 0.20, GROUP)];
        let after = vec![valid_row(true, 0.05, GROUP), valid_row(true, 0.10, GROUP)];
        let result = compare_cost_groups(&before, &after);
        assert!(result.valid);
        assert!(result.cost_delta.is_some());
        let delta = result.cost_delta.unwrap();
        assert!(delta < 0.0);
    }

    #[test]
    fn test_compare_cost_groups_different_groups() {
        let group_a = "suite/fam/v1/crit";
        let group_b = "suite/fam/v2/crit";
        let before = vec![valid_row(true, 0.1, group_a)];
        let after = vec![valid_row(true, 0.1, group_b)];
        let result = compare_cost_groups(&before, &after);
        assert!(!result.valid);
        assert!(result.reason.contains("different cost_of_pass_groups"));
    }

    #[test]
    fn test_compare_cost_groups_no_success() {
        let rows = vec![valid_row(false, 0.1, GROUP)];
        let result = compare_cost_groups(&rows, &rows);
        assert!(!result.valid);
        assert!(result.reason.contains("success_count=0"));
    }

    #[test]
    fn test_detect_invalid_cost_comparison() {
        let group_a = "s/f/v/c1";
        let group_b = "s/f/v/c2";
        let rows_a = vec![valid_row(true, 0.1, group_a)];
        let rows_b = vec![valid_row(true, 0.2, group_b)];
        let (invalid, reason) = detect_invalid_cost_comparison(&rows_a, &rows_b);
        assert!(invalid);
        assert!(reason.contains("different"));

        let rows_same = vec![valid_row(true, 0.1, GROUP)];
        let (invalid2, reason2) = detect_invalid_cost_comparison(&rows_same, &rows_same);
        assert!(!invalid2);
        assert_eq!(reason2, "comparison is valid");
    }

    #[test]
    fn test_usage_ledger_row_to_value() {
        let row_json = valid_row(true, 0.5, GROUP);
        let violations = validate_usage_ledger_row(&row_json);
        assert!(violations.is_empty());
    }
}
